src/functions.c

```diff
@@ -372,59 +375,148 @@ void fcallroCommand(client *c) {
+        return;
+    }
+
+    functionsCtx *f_ctx = functionsCtxCreate();
     rioInitWithBuffer(&payload, data);
 
     /* Read until reaching last 10 bytes that should contain RDB version and checksum. */
-    while(data_len - payload.io.buffer.pos > 10) {
-        /* Read type. */
-        if ((type = rdbLoadType(&payload)) == -1) goto load_error;
-        if (type != RDB_OPCODE_FUNCTION) goto load_error;
-        if (rdbFunctionLoad(&payload, rdbver, f_ctx) != C_OK) goto load_error;
+    while (data_len - payload.io.buffer.pos > 10) {
+        int type;
+        if ((type = rdbLoadType(&payload)) == -1) {
+            err = sdsnew("can not read data type");
+            goto load_error;
+        }
+        if (type != RDB_OPCODE_FUNCTION) {
+            err = sdsnew("given type is not a function");
+            goto load_error;
+        }
+        if (rdbFunctionLoad(&payload, rdbver, f_ctx, &err) != C_OK) {
+            goto load_error;
+        }
     }
 
-    functionsCtxSwapWithCurrent(f_ctx);
-    forceCommandPropagation(c, PROPAGATE_REPL | PROPAGATE_AOF);
-    addReply(c,shared.ok);
-    return;
+    if (restore_replicy == RESTORE_POLICY_FLUSH) {
+        functionsCtxSwapWithCurrent(f_ctx);
+        f_ctx = NULL; /* avoid releasing the f_ctx in the end */
+    } else {
+        if (restore_replicy == RESTORE_POLICY_APPEND) {
+            /* First make sure there is only new functions */
+            iter = dictGetIterator(f_ctx->functions);
+            dictEntry *entry = NULL;
+            while ((entry = dictNext(iter))) {
+                functionInfo *fi = dictGetVal(entry);
+                if (dictFetchValue(functions_ctx->functions, fi->name)) {
+                    /* function already exists, failed the restore. */
+                    err = sdscatfmt(sdsempty(), "Function %s already exists", fi->name);
+                    goto load_error;
+                }
+            }
+            dictReleaseIterator(iter);
+        }
+        iter = dictGetIterator(f_ctx->functions);
+        dictEntry *entry = NULL;
+        while ((entry = dictNext(iter))) {
+            functionInfo *fi = dictGetVal(entry);
+            dictEntry *existing = NULL;
+            dictEntry *new = dictAddRaw(functions_ctx->functions, fi->name,
+                    &existing);
+            if (!new) {
+                functionInfo *old_fi = dictGetVal(existing);
+                engineFunctionDispose(NULL, old_fi);
+                new = existing;
+            }
+            dictSetVal(functions_ctx->functions, new, fi);
+            dictSetVal(f_ctx->functions, entry, NULL); /* make sure value will not be disposed */
+        }
+    }
+
+    /* Indicate that the command changed the data so it will be replicated and
+     * counted as a data change (for persistence configuration) */
+    server.dirty++;
 
 load_error:
-    functionsCtxFree(f_ctx);
-    addReplyError(c,"Failed loading payload");
+    if (err) {
```

Comment:
i don't like the fact that we rely on `err` to be set in order to know we failed.
it would be ok if it's all local to this function, but then we need to make sure to set it manually, if `rdbFunctionLoad` fails and doesn't set it.

Reply:
Thought to spare the extra argument but I can add another argument that indicate failure.

Reply:
or make sure that when rdbFunctionLoad returns an error without setting `err`, you're setting it yourself. (better not trust the implementation of another function)